src/json.ts

```diff
@@ -99,7 +99,10 @@ export type JsonRpcError = OptionalField<
   'data'
 >;
 
-export const JsonRpcParamsStruct = optional(union([object(), array()]));
+export const JsonRpcParamsStruct = optional<
+  Record<string, Json> | Json[],
+  null
+>(union([object(), array()]) as Struct<Record<string, Json> | Json[], null>);
```

Comment:
Wouldn't this work, without the need for a type cast?

```suggestion
export const JsonRpcParamsStruct = optional(
  union([record(string(), JsonStruct), array(JsonStruct)]),
);
```

Reply:
```
Type cast - object x 1,991,153 ops/sec ±0.26% (97 runs sampled)
Type cast - array x 1,763,260 ops/sec ±5.30% (89 runs sampled)
Validation - object x 509,050 ops/sec ±3.49% (86 runs sampled)
Validation - array x 492,139 ops/sec ±0.18% (99 runs sampled)
```

Sure, it's a little bit slower, but I don't think there's any real world difference. And we should validate the parameters anyway, values that aren't JSON-serialisable are not valid JSON-RPC params.

Reply:
3-4 times slower is not something I'd call "little bit slower". `JSON.stringify` tests the parameters for us anyways since it won't serialize otherwise.

Reply:
By the way, what benchmarking tool is that?

Reply:
> 3-4 times slower is not something I'd call "little bit slower".

I mean that it still runs at 500k operations per second. It won't make any significant difference for us.

> `JSON.stringify` tests the parameters for us anyways since it won't serialize otherwise.

Yes, but `JSON.stringify` has a few strange behaviours, like `Infinity` and `NaN` are treated as `null`, `undefined` is completely omitted or replaced with `null` in arrays, etc.

We also don't always use `JSON.stringify`, for example when working with post messaging streams, resulting in bugs as described in MetaMask/snaps-monorepo#840.

> By the way, what benchmarking tool is that?

[Benchmark.js](https://benchmarkjs.com/).